core/lib/scmi-server/scmi_reset_consumer.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <drivers/rstctrl.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <scmi_agent_configuration.h>
+#include <scmi_reset_consumer.h>
+#include <tee_api_defines_extensions.h>
+#include <trace.h>
+
+/*
+ * struct scmi_server_reset: data for a SCMI reset in DT
+ *
+ * @domain_id: SCMI domain identifier
+ * @domain_name: SCMI domain name
+ * @reset: reset to control thru SCMI protocol
+ */
+struct scmi_server_reset {
+	uint32_t domain_id;
+	const char *domain_name;
+	struct rstctrl *reset;
+};
+
+TEE_Result optee_scmi_server_init_resets(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+							*channel_cfg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_server_reset *s_resets = NULL;
+	size_t s_resets_count = 0;
+	int subnode = 0, item_node = 0;
+	bool have_subnodes = false;
+	size_t n = 0;
+
+	item_node = fdt_subnode_offset(fdt, node, "resets");
+	if (item_node < 0)
+		return TEE_SUCCESS;
+
+	/* Compute the number of domains to allocate */
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		paddr_t reg = fdt_reg_base_address(fdt, subnode);
+
+		assert(reg != DT_INFO_INVALID_REG);
+		if ((size_t)reg > s_resets_count)
+			s_resets_count = (uint32_t)reg;
+
+		have_subnodes = true;
+	}
+
+	if (!have_subnodes)
+		return TEE_SUCCESS;
+
+	s_resets_count += 1;
```

Comment:
Maybe worth telling that `/* Number of SCMI reset domains is the max domain ID + 1 */`.